File: dart/detective.dart
import 'diagonalize.dart';
import 'oracle.dart';

class Detective {
  bool verbose;

  Detective({this.verbose:false}) {}

  DetectiveResult search(Oracle oracle, int nGuesses) {
    oracle.reset();
    for (int guess = 0; guess < nGuesses; guess++) {
      DiagonalIndices indices = getCountableIndexes(guess);
      int alphaSecret = indices.v1;
      int deltaSecret = indices.v2;
      int alphaGuess = alphaSecret + guess * deltaSecret;

      if (this.verbose)
        print("Attempt $guess: a0=$alphaSecret, da=$deltaSecret, an=$alphaGuess");
      
      bool guessCorrect = oracle.ask(alphaGuess, deltaSecret);
      if (guessCorrect)
        return DetectiveResult(guess, alphaGuess, alphaSecret, deltaSecret);
    }

    return null;
  }
}

class DetectiveResult {
  int nGuesses, alphaFinal, alphaSecret, deltaSecret;

  DetectiveResult(this.nGuesses, this.alphaFinal, this.alphaSecret, this.deltaSecret) {}
}

File: python/detective.py
import diagonal_search


class Detective:
    def __init__(self, verbose=False):
        self._verbose = verbose

    def search(self, oracle, n_guesses):
        oracle.reset()
        for guess in range(n_guesses):
            alpha_0, delta = diagonal_search.get_countable_indexes(guess)
            alpha_n = alpha_0 + guess * delta

            if self._verbose:
                print(f"Attempt {guess}: a_0={alpha_0}, da={delta}, a_n={alpha_n}")

            guess_result = oracle.ask(alpha_n, delta)
            if guess_result:
                return guess, alpha_0, delta, alpha_n
        return None
